tests/DeadLetterQueueTest.cc

```diff
@@ -0,0 +1,390 @@
+    ASSERT_EQ(consumerImpl.deadLetterPolicy_.getDeadLetterTopic(), dlqTopic);
+
+    client.close();
+}
+
+class DeadLetterQueueTest : public ::testing::TestWithParam<std::tuple<bool, bool, ConsumerType>> {
+   public:
+    void SetUp() override {
+        isProducerBatch_ = std::get<0>(GetParam());
+        isMultiConsumer_ = std::get<1>(GetParam());
+        consumerType_ = std::get<2>(GetParam());
+        producerConf_ = ProducerConfiguration().setBatchingEnabled(isProducerBatch_);
+    }
+
+    void TearDown() override { client_.close(); }
+
+    void initTopic(std::string topicName) {
+        if (isMultiConsumer_) {
+            // call admin api to make it partitioned
+            std::string url = adminUrl + "admin/v2/persistent/public/default/" + topicName + "/partitions";
+            int res = makePutRequest(url, "5");
+            LOG_INFO("res = " << res);
+            ASSERT_FALSE(res != 204 && res != 409);
+        }
+    }
+
+   protected:
+    Client client_{lookupUrl};
+    ProducerConfiguration producerConf_;
+    bool isMultiConsumer_;
+    bool isProducerBatch_;
+    ConsumerType consumerType_;
+};
+
+TEST_P(DeadLetterQueueTest, testSendDLQTriggerByAckTimeOutAndNeAck) {
+    Client client(lookupUrl);
+    const std::string topic = "testSendDLQTriggerByAckTimeOut-" + std::to_string(time(nullptr)) +
+                              std::to_string(isMultiConsumer_) + std::to_string(isProducerBatch_) +
+                              std::to_string(consumerType_);
+    const std::string subName = "dlq-sub";
+    const std::string dlqTopic = topic + "-" + subName + "-DLQ";
+    initTopic(topic);
+
+    auto dlqPolicy = DeadLetterPolicyBuilder().maxRedeliverCount(3).deadLetterTopic(dlqTopic).build();
+    ConsumerConfiguration consumerConfig;
+    consumerConfig.setDeadLetterPolicy(dlqPolicy);
+    consumerConfig.setNegativeAckRedeliveryDelayMs(100);
+    consumerConfig.setConsumerType(consumerType_);
+    Consumer consumer;
+    ASSERT_EQ(ResultOk, client.subscribe(topic, subName, consumerConfig, consumer));
+
+    // Reset the unack timeout
+    long unackTimeOut = 200;
+    if (isMultiConsumer_) {
+        auto multiConsumer = PulsarFriend::getMultiTopicsConsumerImplPtr(consumer);
+        multiConsumer->unAckedMessageTrackerPtr_.reset(new UnAckedMessageTrackerEnabled(
+            unackTimeOut, PulsarFriend::getClientImplPtr(client), *multiConsumer));
+        multiConsumer->consumers_.forEachValue([&client, unackTimeOut](ConsumerImplPtr consumer) {
+            consumer->unAckedMessageTrackerPtr_.reset(new UnAckedMessageTrackerEnabled(
+                unackTimeOut, PulsarFriend::getClientImplPtr(client), *consumer));
+        });
+    } else {
+        auto &consumerImpl = PulsarFriend::getConsumerImpl(consumer);
+        consumerImpl.unAckedMessageTrackerPtr_.reset(new UnAckedMessageTrackerEnabled(
+            unackTimeOut, PulsarFriend::getClientImplPtr(client), consumerImpl));
+    }
```

Comment:
I think a better way to set the `unackTimeout` is to modify the `impl_` field of `ConsumerConfiguration` directly, so we can pass the check in `ConsumerConfiguration::setUnAckedMessagesTimeoutMs`.

Reply:
In this case we need to add `friend` to the `ConsumerConfiguration` include file. This may expose the test class to the user.

Is there any good way?

Reply:
The class won't be exposed. The `PulsarFriend` is already exposed in these headers.

In addition, the friend declaration does not **expose** anything. It just indicates you can implement a class with the same name. Exposing usually means the methods and fields are exposed to users.